### ISSUE TITLE
tlvf - structs are incorrect

### ISSUE THREAD

**Issue by arnout** [prplfoundation/prplMesh#18]
tlvf creates C++ structures and maps them directly into a uint8_t* buffer. This does not work for two reasons:
* it relies on CPU endianness to be the same as the byte stream's;
* it relies on struct packing to be the same as the byte stream's.

The first one is worked around by adding struct_swap(), but the second issue still exists. This leads to additional bytes in the generated stream. E.g. WSC::sM2 starts with
```
typedef struct sM2 {
    WSC::sWscAttrVersion version_attr;
    WSC::sWscAttrMessageType message_type_attr;
   ...
```
The byte stream is generated as `10 4a 00 01 10 00 10 22 00 01 04 00`, with a spurious 0 byte after both attributes.

Note that a similar (but less critical) issue exists for classes, except there it's an unaligned access. For classes, the members are accessed through pointers which are set into the buffer at the correct offset; if that offset is not properly aligned, the members will be accessed through an unaligned pointer. Fortunately, we only target CPU architectures (ARMv7A, ARMv8, x86_64) that support unaligned access, so this is not much of an issue.

There are three solutions to this issue.
1. Add `__attribute__((packed))` to all structs. This is the simplest solution, but feels like a hack. Of course, struct_swap already is a hack so maybe it's not a problem.
2. Use accessors for accessing struct members instead of writing directly to the buffer. This can handle the writes and swapping at the same time. However, it changes the interface (it's no longer possible to write struct members directly).
3. Use the structs as actual structs, and add a finalize function that generates the bytes into a buffer. The finalize function can do the swapping too.

**Comment by tomereli** [comment]
Nice catch! I think we should go with solution 1, which is the simplest.
Maybe in the future, we can rewrite tlvf to work more like protobuf and the likes of it - do the serialization/de-serialization and endianness conversion when finalize/parse are called.
I expect that it will make the generated code much more readable and less error-prone.

**Comment by rammarzin** [comment]
I also vote for option #1

‫בתאריך יום ג׳, 11 ביוני 2019 ב-22:54 מאת ‪Tomer Eliyahu‬‏ <‪
notifications@github.com‬‏>:‬

**Comment by tomereli** [comment]
Fixed in #27
